trie/trie_test.go

```diff
@@ -141,108 +163,109 @@ func TestBranch(t *testing.T) {
-		t.Errorf("Fail to get key %x with value %x: got %x", key3, value3, val)
+	for _, test := range tests {
+		err := trie.Put(test.key, test.value)
+		if err != nil {
+			t.Errorf("Fail to put with key %x and value %x: %s", test.key, test.value, err.Error())
+		}
 	}
 
-	val, err = trie.Get(key4)
-	if err != nil {
-		t.Errorf("Fail to get key %x: %s", key4, err.Error())
-	} else if !bytes.Equal(val, value4) {
-		t.Errorf("Fail to get key %x with value %x: got %x", key4, value4, val)
+	for _, test := range tests {
+		val, err := trie.Get(test.key)
+		if err != nil {
+			t.Errorf("Fail to get key %x: %s", test.key, err.Error())
+		} else if !bytes.Equal(val, test.value) {
+			t.Errorf("Fail to get key %x with value %x: got %x", test.key, test.value, val)
+		}
 	}
 }
 
-func TestBranchMore(t *testing.T) {
-	trie := newEmpty()
-
-	key1 := []byte{0x01}
-	value1 := []byte("spaghetti")
-	key2 := []byte{0x02}
-	value2 := []byte("gnocchi")
-	key3 := []byte{0xf7}
-	value3 := []byte("ramen")
-	key4 := []byte{0x43}
-	value4 := []byte("pho")
-
-	err := trie.Put(key1, value1)
-	if err != nil {
-		t.Errorf("Fail to put with key %x and value %x: %s", key1, value1, err.Error())
-	}
+func TestPutAndGet(t *testing.T) {
+	for i := 0; i < 10; i++ {
+		trie := newEmpty()
+		rt := generateRandTest(1000)
+		for _, test := range rt {
+			err := trie.Put(test.key, test.value)
+			if err != nil {
+				t.Errorf("Fail to put with key %x and value %x: %s", test.key, test.value, err.Error())
+			}
 
-	err = trie.Put(key2, value2)
-	if err != nil {
-		t.Errorf("Fail to put with key %x and value %x: %s", key2, value2, err.Error())
-	}
+			val, err := trie.Get(test.key)
+			if err != nil {
+				t.Errorf("Fail to get key %x: %s", test.key, err.Error())
+			} else if !bytes.Equal(val, test.value) {
+				t.Errorf("Fail to get key %x with value %x: got %x", test.key, test.value, val)
+			}
+		}
 
-	err = trie.Put(key3, value3)
-	if err != nil {
-		t.Errorf("Fail to put with key %x and value %x: %s", key3, value3, err.Error())
-	}
+		for _, test := range rt {
+			val, err := trie.Get(test.key)
+			if err != nil {
+				t.Errorf("Fail to get key %x: %s", test.key, err.Error())
+			} else if !bytes.Equal(val, test.value) {
+				t.Errorf("Fail to get key %x with value %x: got %x", test.key, test.value, val)
 
-	err = trie.Put(key4, value4)
-	if err != nil {
-		t.Errorf("Fail to put with key %x and value %x: %s", key4, value4, err.Error())
-	}
+				tests := ""
+				for _, othertest := range rt {
+					tests = fmt.Sprintf("%s%s\n%s\n", tests, othertest.key, othertest.value)
+				}
 
-	val, err := trie.Get([]byte{0})
-	if err != nil {
-		t.Errorf("Fail to get key %x: %s", key1, err.Error())
-	} else if !bytes.Equal(val, nil) {
-		t.Errorf("Fail to get key %x with nil value: got %x", "noot", val)
+				err := writeToTestFile(tests)
+				if err != nil {
+					t.Error(err)
+				}
+			}
+		}
 	}
+}
 
-	val, err = trie.Get(key1)
+func TestFailingTests(t *testing.T) {
```

Comment:
added a comment to explain. if the random tests fail, it writes the test case to the file `trie/test_data`. then, this test can be run to re-run the failing test by reading it from the file `trie/test_data`

Reply:
Still don't really understand why this would be here because if a test fails it will produce `./test_data` which will then immediately be read in this test. 

The ability to re-run past test cases seems useful but perhaps this should be a regular function not explicitly a test.

Reply:
yeah but if its a test its easier to call, just `go test -run TestFailingTests` if its a function I'm not sure what a good flow for it would be?